api/server/sdk/rest_gateway.go

```diff
@@ -0,0 +1,139 @@
+/*
+Package sdk is the gRPC implementation of the SDK gRPC server
+Copyright 2018 Portworx
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package sdk
+
+import (
+	"context"
+	"fmt"
+	"mime"
+	"net/http"
+
+	"github.com/gobuffalo/packr"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+
+	"github.com/libopenstorage/openstorage/api"
+	"github.com/libopenstorage/openstorage/pkg/grpcserver"
+)
+
+type sdkRestGateway struct {
+	config     ServerConfig
+	restPort   string
+	grpcServer *sdkGrpcServer
+}
+
+func newSdkRestGateway(config *ServerConfig, grpcServer *sdkGrpcServer) (*sdkRestGateway, error) {
+	return &sdkRestGateway{
+		config:     *config,
+		restPort:   config.RestPort,
+		grpcServer: grpcServer,
+	}, nil
+}
+
+func (s *sdkRestGateway) Start() error {
+	mux, err := s.restServerSetupHandlers()
+	if err != nil {
+		return err
+	}
+
+	ready := make(chan bool)
+	go func() {
+		ready <- true
+		var err error
+		address := ":" + s.restPort
+		if s.config.Tls != nil {
+			err = http.ListenAndServeTLS(address, s.config.Tls.CertFile, s.config.Tls.KeyFile, mux)
+		} else {
+			err = http.ListenAndServe(address, mux)
+		}
+
+		if err != nil {
+			logrus.Fatalf("Unable to start SDK REST gRPC Gateway: %s\n",
+				err.Error())
+		}
+	}()
+	<-ready
+	logrus.Infof("SDK gRPC REST Gateway started on port :%s", s.restPort)
+
+	return nil
+
+}
+
```

Comment:
We should also add a Stop routine for the  REST gateway.

Reply:
That's much harder to do. It has to do with http server in golang. If it is supports it in later golang versions, I will add it.

Reply:
This is how we can do it
```suggestion
 restGatewayServer := &http.Server{Addr: address, Handler: max}
 restGatewayServer.ListenAndServeTLS(s.config.Tls.CertFile, s.config.Tls.KeyFile)
 
 
 // for stopping
 restGatewayServer.Close()
 
```

Reply:
I also added use of the rwlock interceptor to make sure no pending requests are in transit